### DISCUSSION TITLE
Problem with French IPTV

### DISCUSSION THREAD

**Discussion by iamnague** [bluenviron/mediamtx#273] [Questions]
## Which version are you using?

v0.13.3 (docker)

## Which operating system are you using?

OS

- [ x] Linux
- [ ] Windows
- [ ] macOS

Architecture

- [x ] amd64
- [ ] arm64
- [ ] arm7
- [ ] arm6

## Describe the issue

My ISP (Free, France) is offering IPTV service based on RSTP/RTP. Rstp-simple-server, in proxy configuration, can't dectect this IPTV source: ERR: no UDP packets received recently (maybe there's a firewall/NAT in between)

[debug_rtsp-simple-server.txt](https://github.com/aler9/rtsp-simple-server/files/5889037/debug_rtsp-simple-server.txt)

This source is UDP/RTP with payload type 33 (MPEG II TS). It works with VLC. 
Rstp-simple-server, same configuration, works in proxy mode with a different RSTP/RTP source (payload type 97 for ex).

## Describe how to replicate the issue

You need this source to replicate. See attached dump with the ko source, and a good one.

## Did you attach a network dump?

[captures.zip](https://github.com/aler9/rtsp-simple-server/files/5889012/captures.zip)

**Comment by aler9** [comment]
Hello, to solve this i need 

* 30 secs of data between VLC and the IPTV (use RTSP, not just RTP)
* 30 secs of data between rtsp-simple-server and the IPTV

Unfortunately your dumps contained only UDP packets, and don't contain the RTSP handshake - we have to find out whether
* there's a different handshake
* NAT is not opened correctly
* there's something else

Furthermore, please try the undockerized version too.

**Comment by iamnague** [comment]
Hello @aler9 , thanks for your help.

Here the requested pcap dumps: [captures2.zip](https://github.com/aler9/rtsp-simple-server/files/5895281/captures2.zip)

I hope it can help you.

**Comment by iamnague** [comment]
Hello @aler9,
FYI, I've tried last release with the undockerized version, on a different host OS (Ubuntu 16.04) same issue:

```
./rtsp-simple-server
2021/02/01 14:34:31 [I] [0/0/0] rtsp-simple-server v0.14.0
2021/02/01 14:34:31 [I] [0/0/0] [UDP/RTP listener] opened on 0.0.0.0:8000
2021/02/01 14:34:31 [I] [0/0/0] [UDP/RTCP listener] opened on 0.0.0.0:8001
2021/02/01 14:34:31 [I] [0/0/0] [TCP/RTSP listener] opened on 0.0.0.0:8554
2021/02/01 14:34:31 [I] [0/0/0] [path proxied1] [rtsp source] started
2021/02/01 14:34:31 [I] [0/0/0] [path proxied1] [rtsp source] connecting
2021/02/01 14:34:31 [I] [0/0/0] [path proxied1] [rtsp source] ready
2021/02/01 14:34:41 [I] [0/0/0] [path proxied1] [rtsp source] ERR: no UDP packets received recently (maybe there's a firewall/NAT in between)
```

**Comment by aler9** [comment]
Hello, what happens is a very strange thing.

The server tells the client that the UDP ports are 32792-32793

```
RTSP/1.0 200 OK
Cseq: 3
Server: fbxrtspd/1.2 Freebox RTSP server
Public: DESCRIBE, OPTIONS, SETUP, TEARDOWN, PLAY, PAUSE
X-Freebox-ANPT-Range: 4004334049/4004334049/4004334049/4004334049
Session: H20gKSxsvfwCX0i;timeout=30
Transport: RTP/AVP;unicast;mode=play;destination=212.27.38.254;client_port=48230-48231;server_port=32792-32793
```

But then the server uses two completely different ports to send packets (48230-48231)...

This happens with both VLC and rtsp-simple-server, but the former doesn't check the sender ports, while the latter does for security reasons.

**Comment by aler9** [comment]
In my opinion you can do 3 things:

* check whether your NAT changes UDP source ports
* if this is not the case, write to the TV provider and tell them to fix their RTSP server to send the right server port
* anyway, use TCP

```yml
paths:
  mypath:
    source: rtsp://
    sourceProtocol: tcp
```

I'm sorry but i don't want to remove the security check of the UDP source ports.

**Comment by iamnague** [comment]
Hello @aler9 
Thank you very much for your help! 
Indeed, I focus my troubleshooting on client_port and I never checked server_port. You are right, server_port doesn't match UDP/RTP source port. And it's not my NAT that changes source ports.

Based on RFC2326:
```
server_port:
        This parameter provides the unicast RTP/RTCP port pair on
        which the server has chosen to receive media data and control
        information.  It is specified as a range, e.g.,
        server_port=3456-3457.
```
My understanding is that these ports are not mandatory used for UDP/RTP source ports... but let's say that this part of the RFC is 
ambiguous.

 I've found a dicussion on the TV provider forum about this issue, based on RFC, they won't change the RTSP behavior :(

Is there any chance that you implement a disable "security check" option ?

Edit: this TV provider doesn't support TCP : ERR: bad status code: 461 (Unsupported Transport)

**Comment by aler9** [comment]
Hello, in my opinion the best solution consists in hacking the code and compiling a custom executable by yourself.

Edit this file, go to this line: https://github.com/aler9/rtsp-simple-server/blob/2b4d22589ca05559afbbce2635ca0a4fa463286b/internal/sourcertsp/source.go#L124

Add 
```go
AnyPortEnable: true,
```

And recompile with Go (go build . ) or Docker (make release)

In the meanwhile, write to the IPTV provider, they're causing problems that can't be addressed by breaking specifications.

**Comment by iamnague** [comment]
Hello, thanks for your answer. I'll give a try to compile the code myself but I'll loose the ability to easily deploy and maintain the app.

What specifications are you reffering to ? According to RFC2326, my IPTV provider is not breaking anything....

[a comment by an automated account is not shown]
